docs/madr/decisions/018-fault-injection-policy.md

```diff
@@ -0,0 +1,230 @@
+resource for managing fault injection.
+
+FaultInjection in Kuma is implemented with Envoy's [fault injection
+support](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter).
+
+## Considered Options
+
+* Create a MeshFaultInjection
+
+## Decision Outcome
+
+Chosen option: create a MeshFaultInjection
+
+## Decision Drivers
+
+- Replace existing policies with new policy-matching compliant resources
+
+## Solution
+
+### Current configuration
+Below is a sample fault injection configuration.
+
+```yaml
+spec:
+  sources:
+    - match:
+        kuma.io/service: frontend_default_svc_80
+        kuma.io/protocol: http
+  destinations:
+    - match:
+        kuma.io/service: backend_default_svc_80
+        kuma.io/protocol: http
+  conf:
+    abort:
+      httpStatus: 500
+      percentage: 50
+    delay:
+      percentage: 50.5
+      value: 5s
+    responseBandwidth:
+      limit: 50 mbps
+      percentage: 50
+```
+
+The configuration translates to the Envoy configuration in Http's filters:
+
+```yaml
+{
+"name": "envoy.filters.http.fault",
+"typed_config": {
+  "@type": "type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault",
+  "delay": {
+  "fixed_delay": "5s",
+  "percentage": {
+    "numerator": 505000,
+    "denominator": "TEN_THOUSAND"
+  }
+  },
+  "abort": {
+  "http_status": 500,
+  "percentage": {
+    "numerator": 50
+  }
+  },
+  "headers": [
+  {
+    "name": "x-kuma-tags",
+    "safe_regex_match": {
+    "google_re2": {},
+    "regex": ".*&kuma.io/protocol=[^&]*http[,&].*&kuma.io/service=[^&]*frontend_default_svc_80[,&].*"
+    }
+  }
+  ],
+  "response_rate_limit": {
+  "fixed_limit": {
+    "limit_kbps": "50000"
+  },
+  "percentage": {
+    "numerator": 50
+  }
+  }
+}
+```
+
+### Specification
+
+Fault injection supports only `HTTP` protocol and is configured in `HttpFilter`. We can use the header `x-kuma-tags` to match the specific source. In the case of `MeshHealthCheck` which doesn't set `x-kuma-tags` we can put `x-kuma-tags` or create `MeshFaultInjection` that allows only `Mesh` in `From` section.
+
+#### Top level
+
+Top-level targetRef can have all available kinds:
+```yaml
+targetRef:
+ kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+ name: ...
+```
+Fault injection is an inbound policy so only `from` should be configured.
+
+#### From level
+
+We can allow configuration from different sources.
+```yaml
+from:
+ - targetRef:
+     kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+     name: ...
+```
+The only issue is `MeshHealthCheck` which doesn't set the `x-kuma-tags` flag and might not match the specific fault injection configuration. We have 2 options:
+
+* set `x-kuma-tags` header by default for `MeshHealthCheck`
+* change the scope of `From` to `Mesh` so the policy will affect all sources
+
+#### Configuration
+
+```yaml
+default:
+  disabled: false
+  abort:
+    httpStatus: 500
+    percentage: 50
+  delay:
+    percentage: 50.5
```

Comment:
@lobkovilya there's something about floats that should be do here right?

Reply:
I don't think so, the common approach should work:
```go
Percentage *float64 `json:"percentage,omitempty"`
```

Reply:
I'm pretty sure Kubernetes throws a warning to not use floats. I think @bartsmykla had this problem

Reply:
True, I've noticed that @bartsmykla was using `uint` with a factor to avoid float. We could use a string and parse it to float.

Reply:
Do we actually need 0.x precision on percentage here? Wouldn't 1%-100% be enough?

Reply:
https://github.com/kumahq/kuma/blob/master/pkg/plugins/policies/meshcircuitbreaker/api/v1alpha1/meshcircuitbreaker.go#L204

But I agree that we can just allow 1-100%

Reply:
We should maybe come up with some general solution for all the policies: MeshTracing, MeshCircuitBreaker are using a different solutions. MeshTracing allows only int 1-100, while MeshCircuitBreaker has a denominator of 1000 in code which you cannot change. We should maybe come up with a solution that is going to be consistent throughout all the policies.

Reply:
We can implement FractionalPercent struct in `common` like this https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/percent.proto#type-v3-fractionalpercent

Reply:
I'd vote for string `"0.5"`. This is user facing setting. What is easier for users?
```
field:
  numerator: 5
  denominator: 10
```
or:
```
field: "0.5"
```

Reply:
String seems easier to understand 👍

Reply:
The only thing with string is yaml being annoying

Reply:
By default `denominator` could be 100, so most of the time user is doing:
```yaml
percentage:
  numerator: 99
```
But strings are also fine

Reply:
We can try to use `resource.Quantity` from k8s/api